tests/e2e/kafkatopic_webhook.go

```diff
@@ -0,0 +1,318 @@
+// Copyright © 2023 Cisco Systems, Inc. and/or its affiliates
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"strings"
+	"text/template"
+
+	"github.com/Masterminds/sprig"
+	"github.com/gruntwork-io/terratest/modules/k8s"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+// TODO(mihalexa): move to k8s.go
+const (
+	dryRunStrategyNone   string = "none"
+	dryRunStrategyClient string = "client"
+	dryRunStrategyServer string = "server"
+)
+
+// TODO(mihalexa): move to k8s.go
+// applyK8sResourceFromTemplateWithDryRun is copy of applyK8sResourceFromTemplate which calls a "--dry-run=<strategy>" kubectl command
```

Comment:
I feel like this could be solved with just adding an option for dry-run to the original command.

Reply:
Such an option would need to become an input parameter and modifying the initial function was overkill, in my mind, as it's plenty good the way it is and our use of it doesn't normally (aka 90+% of the time) need it to execute with `--dry-run` in mind.  
Outside of what I deemed to be an optimized workflow for webhook testing (aka making use of the dry-run option), in our e2e tests we always want to fully create objects (aka have them written to `etcd`) because that's in the nature of the tests: we create objects and we check their functionality.

The initial function is also part of a chain - it is not really standalone. Modifying its signature means modifying the signatures of all functions that relate to it in any way.

I preferred to add it as an optional/alternative function.  

I could be convinced of the opposite but right now I don't feel like it is a 50-50 choice.

Reply:
Yeah, I get that but at he same time I'd like to avoid major code duplication if we can, and that in my head is more important than what you've written here but I won't hold up the PR on this if everyone else is fine with this reasoning, it's a personal preference thing for me.

Reply:
Yeah, that's a difficult topic.
I definitely wouldn't add a `dryRunStrategy string` parameter to the original function.

But maybe we should pay the one time cost and add an `extraArgs map[string][]string` parameter to that one which is passed down to the TerraTest call through the other helpers.
I know it's a big refactor and I can live with postponing it, but that looks like the desirable ideal solution to me.

Reply:
The chain in question is : applyK8sResourceFromTemplate -> applyK8sResourceManifestFromString -> now moving into `terratest`'s `k8s` pkg -> KubectlApplyFromStringE -> KubectlApplyE (inserts "apply -f") -> RunKubectlE -> RunKubectlAndGetOutputE -> shell.RunCommandAndGetOutputE.

Part of the issue with this chain is that if we want to use `terratest`'s `KubectlApply<...>` functions ([link here](https://github.com/gruntwork-io/terratest/blob/9e475c5822a5ad4a8a09613fb44a4fd61976b6a9/modules/k8s/kubectl.go#L87-L124)), they don't have `...args` in their signature at all - that's the root of the problem and otherwise it would've probably been an easy refactor.  The first point in the chain where we get `...args` is `RunKubectlE`.
Examples of function signatures relevant to us in the current implementation (also check comments I added at the end of some lines):
```
func KubectlApplyFromStringE(t testing.TestingT, options *KubectlOptions, configData string) error { 
  ... 
  return KubectlApplyE(t, options, tmpfile)
}
func KubectlApplyE(t testing.TestingT, options *KubectlOptions, configPath string) error {
	return RunKubectlE(t, options, "apply", "-f", configPath)    <<<<< first args...  ;  this is where "apply -f" gets added
}
func RunKubectlE(t testing.TestingT, options *KubectlOptions, args ...string) error {    <<<<< first args...
	_, err := RunKubectlAndGetOutputE(t, options, args...)
	return err
}
```

It is from the implementations behind that chain of functions that I pieced together a direct path to my goal.

Reply:
Do we want to take Marton's suggestion from https://github.com/banzaicloud/koperator/pull/1031#issuecomment-1680492691 (which I am guessing was meant for this thread) and refactor `applyK8sResourceManifestFromString` (and upwards one level in our own functions) to approximate the `KubectlApply...` from terratest instead of using it directly ?  (and in the process eliminate the need for a new function for dry-run ?)

Reply:
> Do we want to take Marton's suggestion from https://github.com/banzaicloud/koperator/pull/1031#issuecomment-1680492691 (which I am guessing was meant for this thread) and refactor applyK8sResourceManifestFromString (and upwards one level in our own functions) to approximate the KubectlApply... from terratest instead of using it directly ? (and in the process eliminate the need for a new function for dry-run ?)

Yeah, I think that was kind of my point as well, I meant it somewhat that way, I see why my thought was simpler but not working, but the end result seems similar to me.